File: algorithms/N/Number of People Aware of a Secret/Number of People Aware of a Secret.js
/**
 * @param {number} n
 * @param {number} delay
 * @param {number} forget
 * @return {number}
 */
var peopleAwareOfSecret = function(n, delay, forget) {
    const dp=new Array(n+1).fill(0);
    let numberOfPeopleSharingSecret = 0;
    let totalNumberOfPeopleWithSecret = 0;
    const MOD = 1000000007n;
    
    dp[1]=1; // as on day one only one person knows the secret
    
    for(let i=2;i<=n;i++){
        const numberOfNewPeopleSharingSecret = dp[Math.max(i-delay,0)];
        const numberOfPeopleForgettingSecret = dp[Math.max(i-forget,0)];
        numberOfPeopleSharingSecret = BigInt(numberOfPeopleSharingSecret) + 
            (   BigInt(numberOfNewPeopleSharingSecret) 
              - BigInt(numberOfPeopleForgettingSecret) 
              + BigInt(MOD)
            ) % BigInt(MOD);
        
        dp[i] = numberOfPeopleSharingSecret;
    }
    
    for(let i=n-forget+1;i<=n;i++){
       totalNumberOfPeopleWithSecret = 
           (BigInt(totalNumberOfPeopleWithSecret) + BigInt(dp[i])) % BigInt(MOD); 
    }
    
    return totalNumberOfPeopleWithSecret;
};

File: scripts/algorithms/N/Number of People Aware of a Secret/Number of People Aware of a Secret.cpp
static int MOD=1e9+7;
class Solution {
public:
    int delay,forget;
    vector<long> memo;
    // Total number of people who would have found out about the secret by the nth day.
    long dp(int n) {
        if(!n)
            return 0;
        if(memo[n]!=-1)  // Return cached result if exists.
            return memo[n];
		// Current contribution of 1 person who knows the secret
        long result=1;
        for(int i=delay;i<forget;i++)       // Number of people that the secret will be forwarded to
            if(n-i>=0)
                result=(result+dp(n-i))%MOD;
        return memo[n]=result;
    }
    int peopleAwareOfSecret(int n, int delay, int forget) {
        this->delay=delay;
        this->forget=forget;
        memo.resize(n+1,-1);
        return (dp(n)-dp(n-forget)+MOD)%MOD;    // Subtract the people who found out by the `n-forget` day as observed.
    }
};
